refactor: try to optimize redis calls

akhilnarang opened the pull request:
Use pipelines wherever possible, bunch calls wherever possible

- refactor: pipeline wherever possible, optimize calls
- chore: don't delete keys that aren't used

Resolves #24604

akhilnarang committed:
refactor: pipeline wherever possible, optimize calls

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
chore: don't delete keys that aren't used

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

ankush commented on frappe/utils/redis_wrapper.py:
```diff
@@ -238,22 +238,65 @@ def hget(self, name, key, generator=None, shared=False):
 
 	def hdel_keys(self, name_starts_with, key):
 		"""Delete hash names with wildcard `*` and key"""
+		pipeline = self.pipeline()
```
also look at delete_keys/delete_values (used by frappe.clear_cache)

Write tests to ensure these changes with pipeline work. 

`test_caching.py` has (very few) tests for rediswrapper

akhilnarang committed:
refactor(twofactor): reduce number of calls to redis

Pipeline wherever possible, set expiry while saving value

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
refactor(boot): frappe.cache() -> frappe.cache

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
refactor(redis): DEL -> UNLINK

This is beneficial since unlink is non blocking, it will run in a different thread if the data passed is large
https://redis.io/commands/unlink/

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
chore(utils/caching): drop unnecessary indent

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
chore(cache): add in some tests

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix(redis_wrapper/hdel): don't pipeline if we can avoid it

For a single deletion it doesn't make much sense, so just directly delete and return

Signed-off-by: Akhil Narang <me@akhilnarang.dev>
